lib/util/semver.js

```diff
@@ -0,0 +1,483 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+/** @typedef {(string|number|undefined|[])[]} SemVerRange */
+/** @typedef {(string|number|undefined|[])[]} SemVerVersion */
```

Comment:
just curious why not to use
```
/** @typedef {Array<Array<string|number>>} SemVerVersion */
```
It simplifies things a little bit (probably comes with a structure size trade of)
e.g. `versionToString` ie9 compatible
```
versionToString = version => version.reduce((acc, part, i) => acc + (i > 1 ? "+" : i > 0 ? "-" : "") + part.join("."), "")
```

Maybe it is not suitable for `SemVerRange`.. since then we need to use a fixed side `[[1], [], [], [2], [], []]`

Reply:
Interesting thing is that you can compare arrays =)
`[1, 1, 1] < [2] === true`
`["beta", 20] < ["beta", 4] === true`
`["beta"] < ["betaa"] === true`

Reply:
Interesting idea. That's worth trying. I think I refactor the whole thing in a way to not leak the parsed version into the API, so all this is webpack internal and can be changed in future.